main/influxdb_watcher.go

```
@@ -0,0 +1,205 @@
+	PrometheusEndpoint string `env:"PROMETHEUS_ENDPOINT" envDefault:"0.0.0.0:8080"`
+	Period             int64  `env:"WATCHER_PERIOD" envDefault:"600"`
+	Timeout            int64  `env:"WATCHER_TIMEOUT" envDefault:"60"`
+}
+
+// InfluxdbBroker Watcher for InfluxDB
+type InfluxdbBroker struct {
+	MonascaDB  string
+	Connection client.Client
+}
+
+// Message the values Watcher is expecting to send and receive
+type Message struct {
+	UUID     string
+	SentTime string
+}
+
+// WriteMessage Write a point to Influxdb with UUID and SentTime
+func (broker *InfluxdbBroker) WriteMessage(byteMessage []byte) error {
+	var message Message
+	err := json.Unmarshal(byteMessage, &message)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Create a new point batch
+	bp, err := client.NewBatchPoints(client.BatchPointsConfig{
+		Database:  broker.MonascaDB,
+		Precision: "s",
+	})
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+
+	// Create a point and add to batch
+	tags := map[string]string{"watcher": "influxdb"}
+	fields := map[string]interface{}{
+		"uuid":      message.UUID,
+		"startTime": message.SentTime,
+	}
+
+	pt, err := client.NewPoint("watcher.influxdb", tags, fields, time.Now())
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+
+	bp.AddPoint(pt)
+
+	// Write the batch
+	if err := broker.Connection.Write(bp); err != nil {
+		log.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+// ReadMessage query InfluxDB for latest measurement
+func (broker *InfluxdbBroker) ReadMessage(timeout time.Duration) (*[]byte, error) {
+	cmd := fmt.Sprintf("SELECT last(startTime), last(uuid) FROM \"%s\"", "watcher.influxdb")
+	res, err := queryDB(broker, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	message := Message{
+		UUID:     res[0].Series[0].Values[0][2].(string),
+		SentTime: res[0].Series[0].Values[0][1].(string),
+	}
+	bytes, err := json.Marshal(message)
+	if err != nil {
+		panic("json encoding failed")
```

This will also kill the watcher. You should instead return the err

---

Done